casper/src/test/scala/coop/rchain/casper/blocks/proposer/BlockReceiverSpec.scala

```diff
@@ -0,0 +1,161 @@
+      validatedBlocksStream = validatedBlocksQueue.dequeue
+      br                    <- BlockReceiver(state, incomingBlockStream, validatedBlocksStream, shardId)
+      res                   <- f(incomingBlockQueue, validatedBlocksQueue, br, node)
+    } yield res
+  }
+
+  private def makeDeploy =
+    ConstructDeploy.sourceDeployNowF("new x in { x!(0) }", shardId = genesis.genesisBlock.shardId)
+
+  private def makeBlock(
+      node: TestNode[Task],
+      justifications: Seq[BlockHash] = Seq()
+  ): Task[BlockMessage] =
+    for {
+      deploy <- makeDeploy
+      block <- node
+                .createBlockUnsafe(deploy)
+                .map(
+                  _.copy(
+                    justifications = justifications.map(Justification(ByteString.EMPTY, _)).toList
+                  )
+                )
+      (sk, _)     = Secp256k1.newKeyPair
+      validatorId = ValidatorIdentity(sk)
+    } yield validatorId.signBlock(block)
+
+  private def addBlock(node: TestNode[Task]): Task[BlockMessage] = makeDeploy >>= (node.addBlock(_))
+
+  it should "pass correct block to output stream" in {
+    withBlockReceiverEnv("root") {
+      case (incomingQueue, _, outStream, node) =>
+        for {
+          block   <- makeBlock(node)
+          _       <- incomingQueue.enqueue1(block)
+          outList <- outStream.take(1).compile.toList
+        } yield outList.length shouldBe 1
+    }
+  }
+
+  it should "discard block with invalid shard name" in {
+    // Provided to BlockReceiver shard name ("test") is differ from block's shard name ("root" by default)
+    // So block should be rejected and output stream should never take block
+    withBlockReceiverEnv("test") {
+      case (incomingQueue, _, outStream, node) =>
+        for {
+          block <- makeBlock(node)
+          _     <- incomingQueue.enqueue1(block)
+        } yield outStream should notEmit
+    }
+  }
+
+  it should "discard block with invalid block hash" in {
+    withBlockReceiverEnv("root") {
+      case (incomingQueue, _, outStream, node) =>
+        for {
+          block <- makeBlock(node).map(_.copy(blockHash = "abc".unsafeHexToByteString))
+          _     <- incomingQueue.enqueue1(block)
+        } yield outStream should notEmit
+    }
+  }
+
+  it should "discard block with invalid signature" in {
+    withBlockReceiverEnv("root") {
+      case (incomingQueue, _, outStream, node) =>
+        for {
+          block <- makeBlock(node).map(_.copy(sig = "abc".unsafeHexToByteString))
+          _     <- incomingQueue.enqueue1(block)
+        } yield outStream should notEmit
+    }
+  }
+
+  it should "discard known block" in {
+    withBlockReceiverEnv("root") {
+      case (incomingQueue, _, outStream, node) =>
+        for {
+          block <- addBlock(node)
+          _     <- incomingQueue.enqueue1(block)
+        } yield outStream should notEmit
+    }
+  }
+
+  "BlockReceiver" should "pass to output blocks with resolved dependencies" in {
```

Comment:
I noticed that in some cases tests can be skipped if `it` is used on the very first test instead of string description.

`it` has a meaning to reference the first test defined previously with the _string_ description, like nested tests. So to prevent any ambiguity let's always define at least first test with string description.